lib/nx_gzlib.c

```diff
@@ -173,51 +203,80 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
 	return file;
 }
 
-int nx_gzread(gzFile file, void *buf, unsigned len)
+gzFile nx_gzopen_(const char* path, int fd, const char *mode) {
+	return __gzopen_(path, fd, mode, 1);
+}
+
+gzFile nx_gzdopen(int fd, const char *mode)
+{
+	return __gzopen_(NULL, fd, mode, 1);
+}
+
+gzFile nx_gzopen(const char* path, const char *mode)
+{
+	return __gzopen_(path, 0, mode, 1);
+}
+
+int __gzread(gzFile file, void *buf, unsigned len, int force_nx)
 {
 	gz_statep state = (gz_statep) file;
 	z_streamp stream = &(state->strm);
 	int ret, err;
 	const uInt max = len > 10? 10 : 1;
 	uLong last_total_out = stream->total_out;
-	unsigned char* next_in = malloc(sizeof(char)*max);
+	typeof(inflate) (*inf);
 
-	if (next_in == NULL)
-		return Z_MEM_ERROR;
+	if (force_nx)
+		inf = &nx_inflate;
+	else
+		inf = &inflate;
 
 	stream->next_out = buf;
 	stream->avail_out = len;
 	do {
-		ret = read(state->fd, next_in, sizeof(char)*max);
-		stream->next_in = next_in;
+		if (state->used_buf == 0) {
+			ret = read(state->fd, state->buf, max);
+			stream->next_in = state->buf;
+		} else {
+			/* There are still data pending in the buffer. */
+			ret = state->used_buf;
+			stream->next_in = state->cur_buf;
+		}
 		if (ret == 0){
 			/* Flush all at the end of the file. */
-			err = nx_inflate(stream, Z_FINISH);
+			err = (*inf)(stream, Z_FINISH);
 			break;
 		}
 
 		stream->avail_in = ret;
-		err = nx_inflate(stream, Z_NO_FLUSH);
+		err = (*inf)(stream, Z_NO_FLUSH);
 
+		state->used_buf = stream->avail_in;
+		state->cur_buf  = stream->next_in;
 		if (err != Z_OK && err != Z_STREAM_END){
 			state->err = err;
-			free(next_in);
 			return 0;
 		}
 	} while (stream->avail_out != 0);
 
-	/* gzread should return when len bytes are wrote on buf, if we still
-	   have unprocessed data on avail_in we need to save it on fifo_in.  */
-	if (stream->avail_in > 0){
-		nx_streamp s = (nx_streamp) stream->state;
-		copy_data_to_fifo_in(s);
-	}
```

Comment:
I didn't get why this block isn't needed anymore. Even though the loop may process leftover input bytes from previous inflate calls, the code exits from it if there's no output space left, which can still happen now.

Reply:
@mscastanho That's because we are not freeing/allocating the input buffer between `__gzread()` calls.
If we leave because it needs more output space, all the input data will be kept at `state->cur_buf` and we can just call `inflate()` again the next time `__gzread()` is called.